containers/sprm-to-json/context/main.py

```diff
@@ -0,0 +1,155 @@
+    cells = write_genes_or_factors_to_cells(
+        input_file=genes_input, cells=cells, genes=True
+    )
+    # Write factors to cells file.
+    cluster_input = Path(input_dir) / Path(tile_str + CLUSTER_FILE_SUFFIX)
+    cells = write_genes_or_factors_to_cells(
+        input_file=cluster_input, cells=cells, genes=False
+    )
+    with open(Path(output_dir) / Path(tile_str).with_suffix(".cells.json"), "w") as f:
+        f.write(json.dumps(cells))
+
+
+def create_factors(tile_str, input_dir, output_dir):
+    factors = {}
+    cluster_input = Path(input_dir) / Path(tile_str + CLUSTER_FILE_SUFFIX)
+    df = pd.read_csv(cluster_input).set_index("ID")
+    cluster_types = df.columns.values
+    df_items = sprm_to_items(cluster_input)
+    for cluster_type in cluster_types:
+        cluster_names = sorted(df[cluster_type].unique().astype("uint8"))
+        factors[cluster_type] = {
+            "map": [str(cluster) for cluster in cluster_names],
+            "cells": {k: v[cluster_type] for (k, v) in df_items},
+        }
+    with open(Path(output_dir) / Path(tile_str).with_suffix(".factors.json"), "w") as f:
+        f.write(json.dumps(factors))
+
+
+def create_genes(tile_str, input_dir, output_dir):
+    genes = {}
+    genes_input = Path(input_dir) / Path(tile_str + GENES_FILE_SUFFIX)
+    df = pd.read_csv(genes_input).set_index("ID")
+    gene_types = df.columns.values
+    df_items = sprm_to_items(genes_input)
+    for gene_type in gene_types:
+        genes[gene_type] = {
+            "max": max(df[gene_type]),
+            "cells": {k: v[gene_type] for (k, v) in df_items},
+        }
+    with open(Path(output_dir) / Path(tile_str).with_suffix(".genes.json"), "w") as f:
+        f.write(json.dumps(genes))
+
+
+def create_clusters(tile_str, input_dir, output_dir):
+    clusters = {}
+    genes_input = Path(input_dir) / Path(tile_str + GENES_FILE_SUFFIX)
+    df = pd.read_csv(genes_input).set_index("ID")
+    gene_types = df.columns.values
+    cell_ids = df.index.values
+    clusters["rows"] = [str(gene_type) for gene_type in gene_types]
+    clusters["cols"] = [str(cell_id) for cell_id in cell_ids]
+    clusters["matrix"] = []
+    for gene_type in gene_types:
+        gene_col = df[gene_type]
+        max_min_diff = max(gene_col) - min(gene_col)
+        # If max_min difference is 0, just output a column of 0's.
+        normalized = (
+            (gene_col - min(gene_col)) / max_min_diff
+            if max_min_diff != 0
+            else (gene_col - gene_col)
+        )
+        clusters["matrix"].append(list(normalized.values))
+    with open(
+        Path(output_dir) / Path(tile_str).with_suffix(".clusters.json"), "w"
+    ) as f:
+        f.write(json.dumps(clusters))
+
+
+def main(input_dir, output_dir):
+    makedirs(output_dir, exist_ok=True)
+    for input_file in glob(input_dir + "/*.ome.tiff-cell_polygons_spatial.csv"):
+        tile_str = Path(input_file).name.split(".")[0]
```

Comment:
```suggestion
        tile_str = Path(input_file).stem
```
... but not quite the same if there are multiple periods in the name.

Reply:
There are basically guaranteed to be many.